internal/agent.go

```diff
@@ -80,38 +79,52 @@ func getAgentID() (string, error) {
 }
 
 // Start the Agent. This will start the discovery ticker and the heartbeat ticker
-func (a *Agent) Start() error {
-	var wg sync.WaitGroup
+func (a *Agent) Start(ctx context.Context) error {
+	g, ctx := errgroup.WithContext(ctx)
 
 	for _, d := range a.discoveries {
-		wg.Add(1)
-		go func(wg *sync.WaitGroup, d discovery.Discovery) {
-			log.Infof("Starting %s loop...", d.GetId())
-			defer wg.Done()
-			a.startDiscoverTicker(d)
-			log.Infof("%s discover loop stopped.", d.GetId())
-		}(&wg, d)
+		dLoop := d
+		g.Go(func() error {
+			log.Infof("Starting %s loop...", dLoop.GetId())
+			a.startDiscoverTicker(ctx, dLoop)
+			log.Infof("%s discover loop stopped.", dLoop.GetId())
+			return nil
+		})
 	}
 
-	wg.Add(1)
-	go func(wg *sync.WaitGroup) {
+	g.Go(func() error {
 		log.Info("Starting heartbeat loop...")
-		defer wg.Done()
-		a.startHeartbeatTicker()
+		a.startHeartbeatTicker(ctx)
 		log.Info("heartbeat loop stopped.")
-	}(&wg)
-
-	wg.Wait()
+		return nil
+	})
+
+	if a.config.FactsEngineEnabled {
+		c := factsengine.NewFactsEngine(a.agentID, a.config.FactsServiceUrl)
+		g.Go(func() error {
+			log.Info("Starting fact gathering service...")
+			if err := c.Subscribe(); err != nil {
+				return err
+			}
+
+			if err := c.Listen(ctx); err != nil {
+				return err
+			}
+
+			log.Info("fact gathering stopped.")
+			return nil
+		})
+	}
 
-	return nil
+	return g.Wait()
```

Comment:
If we want to be more clear, we can wait for the errgroup, and add an if to check the error, and errors.Is if the error is context canceled, so we can log and maybe later distinguish when the error is context canceled (most of the times is a legit operation done by us) and when the error is unexpected

example 

```go
err = g.Wait()
if err != nil {
 if errors.Is(err, context.Canceled){
  // our fault? log this specific case
}
// anyway return the err
return err
}
// errgroup done without error
return nil
```

Reply:
Perfect, we can track it and we can deal with it later :D